### PR TITLE
Add guidance for using qualified names in packages

### PR DESCRIPTION
**Pull request by MasonProtter**
Note: this PR was merged into #42080. It was NOT merged into `master`.

---

Note: the base branch (target branch) of this PR is `kc/warn_using` (#42080), NOT `master`.

---

This is additive to https://github.com/JuliaLang/julia/pull/42080 which adds advice to the docstring of  `using` .

The wording here is a tweaked version of @IanButterworth's suggestion here: https://github.com/JuliaLang/julia/pull/42080#discussion_r1340812119

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Can/should we merge this into https://github.com/JuliaLang/julia/pull/42080?

**Comment by MasonProtter** [comment]
If we think that would help get either of them merged faster then sure. @KristofferC, do you want to merge this branch into yours? Or should we just keep them as separate PRs?

File: doc/src/manual/modules.md [hunk 1]
```diff
@@ -144,8 +144,8 @@ In contrast,
 julia> import .NiceStuff
 ```
 
-brings *only* the module name into scope. Users would need to use `NiceStuff.DOG`, `NiceStuff.Dog`, and `NiceStuff.nice` to access its contents. Usually, `import ModuleName` is used in contexts when the user wants to keep the namespace clean.
-As we will see in the next section `import .NiceStuff` is equivalent to `using .NiceStuff: NiceStuff`.
+brings *only* the module name into scope. Users would need to use `NiceStuff.DOG`, `NiceStuff.Dog`, and `NiceStuff.nice` to access its contents.
+As we will see in the next section `import .NiceStuff` is equivalent to `using .NiceStuff: NiceStuff`. Usually, `import ModuleName` or `using ModuleName: ModuleName` is used in contexts when the user wants to keep the namespace clean.
```
Comment (fingolfin) [inline doc/src/manual/modules.md:148]: Why the sudden switch from `NiceStuff` to `ModuleName` ? I am fine with either name, but shouldn't it be used consistently, i.e. either everything with `NiceStuff` or else everything with `ModuleName`?

File: doc/src/manual/modules.md [hunk 2]
```diff
@@ -168,6 +168,23 @@ Importantly, the module name `NiceStuff` will *not* be in the namespace. If you
 julia> using .NiceStuff: nice, DOG, NiceStuff
 ```
 
+!!! note
+    Qualifying the names being used as in `using NiceStuff: NiceStuff, nice` is recommended over plain
+    `using Foo` for released packages, and other code which is meant to be re-used in the future with
```
Comment (fingolfin) [inline doc/src/manual/modules.md:173]: A third name, `Foo`... Let's stick with one, e.g. `NiceStuff` perhaps?
```suggestion
    `using NiceStuff` for released packages, and other code which is meant to be re-used in the future with
```

File: doc/src/manual/modules.md [hunk 3]
```diff
@@ -168,6 +168,23 @@ Importantly, the module name `NiceStuff` will *not* be in the namespace. If you
...
+    updated dependencies or future versions of julia.
+
+    The reason for this is if another dependency starts to export one of the
+    same names as `Foo` the code will error due to an ambiguity in which
```
Comment (fingolfin) [inline doc/src/manual/modules.md:177]:
```suggestion
    same names as `NiceStuff` the code will error due to an ambiguity in which
```

File: doc/src/manual/modules.md [hunk 4]
```diff
@@ -168,6 +168,23 @@ Importantly, the module name `NiceStuff` will *not* be in the namespace. If you
...
+    package the name should be taken from. This is especially problematic in
+    released packages which needs to be forward-compatible with the future
+    releases (e.g., when `Foo = "1"` and `Bar = 2` are listed in `[deps]`
```
Comment (fingolfin) [inline doc/src/manual/modules.md:180]:
```suggestion
    released packages which should be forward-compatible with future releases
    of their dependencies. For example, when `Foo = "1"` and `Bar = 2` are listed in `[deps]`
```

File: doc/src/manual/modules.md [hunk 5]
```diff
@@ -168,6 +168,23 @@ Importantly, the module name `NiceStuff` will *not* be in the namespace. If you
...
+    section of the `Project.toml` of your package, it should be compatible
+    with the future releases of Foo 1.x and Bar 2.y which both export a
+    function `baz`).
```
Comment (fingolfin) [inline doc/src/manual/modules.md:183]: If I didn't already know what this is about, I am not sure I would understand. I think the point you are trying to make is: suppose you are doing `using Foo, Bar`, and it works. But then `Foo` 1.1 exports a new symbol `baz`, while `Bar` already exports a different `baz`. Then suddenly `using Foo, Bar` fails.

File: doc/src/manual/modules.md (hunk 1 above)
Comment (MasonProtter) [reply doc/src/manual/modules.md:148]: That wasn't really my choice, I was just modifying the existing docs with its existing name scheme. See: https://docs.julialang.org/en/v1/manual/modules/#Standalone-using-and-import

FWIW I think maybe the that *could* be overhauled, but I also don't think it's necessary.

File: doc/src/manual/modules.md (hunk 2 above)
Comment (MasonProtter) [reply doc/src/manual/modules.md:173]: Right, yeah this naming conflict should be avoided. I think for the purposes of this `!!! note` though, maybe it'd be better to use a new set of names `Foo` and `Bar` because we want to talk about conflicts, and we also want this `!!! note` section to be readable standalone.

**Comment by MasonProtter** [comment]
@fingolfin I've changed the `!!! note` section to only talk about `Foo` and `Bar`. You might be right that it's better to just stick with `NiceStuff` here but I think that's not ideal because it'd be good if this note was understandable on its own, but the `NiceStuff` example has some baggage associated with it like being defined inside `Main` and so we'd need to write `using .NiceStuff` instead of `using NiceStuff`. 

I don't have a super strong opinion here, so let me know what you think of this change.

**Review by ararslan — DISMISSED** [review]

File: doc/src/manual/modules.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (MasonProtter) [reply doc/src/manual/modules.md:148]: This isn't really material to this PR. I was just moving around existing text and adding that there's the option to do `using ModuleName: ModuleName` as well.

File: doc/src/manual/modules.md [hunk 6]
```diff
@@ -168,6 +168,26 @@ Importantly, the module name `NiceStuff` will *not* be in the namespace. If you
...
 ```
 
+!!! note
+    Qualifying the names being used as in `using Foo: Foo, f` is
+    recommended over plain `using Foo` for released packages, and other
+    code which is meant to be re-used in the future with updated dependencies
+    or future versions of julia.
+
+    The reason for this is if another dependency starts to export one of the
+    same names as `Foo` the code will error due to an ambiguity in which
```
[the remark this answers is not in the document]
Comment (ararslan) [reply doc/src/manual/modules.md:178]: I think that's an important detail that shouldn't be omitted, otherwise the situation sounds scarier than it is.

File: doc/src/manual/modules.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ararslan) [reply doc/src/manual/modules.md:148]: I'm not suggesting it be added, I'm challenging the wording used, as the motivation as written isn't particularly useful.

[the remark this answers is not in the document]
Comment (MasonProtter) [reply doc/src/manual/modules.md:148]: Again though, that's not something new in this PR.

File: doc/src/manual/modules.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (MasonProtter) [reply doc/src/manual/modules.md]: done

**Review by mbauman — APPROVED** [review]

**Review by fingolfin — APPROVED** [review]

File: doc/src/manual/modules.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fingolfin) [reply doc/src/manual/modules.md:148]: I agree the "clean" bit is not nice -- but it was already there before. It seems unfair to hold this PR hostage over the author not "improving even more things".

Perhaps I am misunderstanding what you'd wish to have changed, @ararslan, in that case I'd appreciate if you could clarify what changes you are requesting exactly.

[the remark this answers is not in the document]
Comment (DilumAluthge) [reply doc/src/manual/modules.md:148]: @ararslan If this is the only part of the diff that you have concerns about, then we can revert it from the PR and make this change in a separate PR, just in the interest of getting this PR in?

cc: @MasonProtter

**Review by DilumAluthge — APPROVED** [review]

File: doc/src/manual/modules.md (hunk 1 above)
Comment (DilumAluthge) [inline doc/src/manual/modules.md:148]: The code suggestion here is to revert this specific change (because it was a source of confusion during code review), and move it to a separate PR.

```suggestion
brings *only* the module name into scope. Users would need to use `NiceStuff.DOG`, `NiceStuff.Dog`, and `NiceStuff.nice` to access its contents. Usually, `import ModuleName` is used in contexts when the user wants to keep the namespace clean.
As we will see in the next section `import .NiceStuff` is equivalent to `using .NiceStuff: NiceStuff`.
```

[the remark this answers is not in the document]
Comment (DilumAluthge) [reply doc/src/manual/modules.md]: So, in the interest of moving this PR forward, I have removed this specific small change from this PR in https://github.com/JuliaLang/julia/commit/862a5466ba19acee25d566488431123fe5494ca1 and moved it to a new PR (#54575).

**Comment by DilumAluthge** [comment]
I've been giving some more thought as to whether this PR should be merged into #42080 (as opposed to merging this PR directly into `master`).

My guess is that if someone has concerns or objections to #42080, then they'll probably have those same concerns/objections to this PR. So it probably makes sense to merge this PR into #42080, so that the main discussion can occur in #42080 (which is where the majority of the discussion has been happening thus far).

Also, I don't want to give the impression that merging this PR into `master` would be doing a kind of "end run" around the PR review happening in #42080.

Therefore, I propose that we change the target branch (base branch) of this PR (from `master` to `kc/warn_using`) and then merge this PR into #42080.

I'll wait a little bit, and then if there are no objections to that plan, I'll perform those steps.

cc: @KristofferC @mbauman

**Comment by DilumAluthge** [comment]
Alright, hearing no objections, I'll merge this PR into #42080 (instead of into `master`).

**Comment by DilumAluthge** [comment]
I've merged this PR into #42080.
